theano/tensor/opt.py

```diff
@@ -2591,6 +2595,139 @@ def local_adv_sub1_adv_inc_sub1(node):
     return [T.cast(y, node.outputs[0].dtype)]
 
 
+@register_specialize
+@register_stabilize
+@register_canonicalize
+@gof.local_optimizer([IncSubtensor,
+                      AdvancedIncSubtensor,
+                      AdvancedIncSubtensor1])
+def local_useless_incsubtensor_alloc(node):
+    """
+    Replaces an [Advanced]IncSubtensor[1], whose increment is an `alloc` of
+    a fully or partially broadcastable variable, by one that skips the
+    intermediate `alloc` where possible.
+    """
+    if isinstance(node.op, (IncSubtensor,
+                            AdvancedIncSubtensor,
+                            AdvancedIncSubtensor1)):
+        x = node.inputs[0]
+        y = node.inputs[1]
+        i = node.inputs[2:]
+
+        if y.owner is not None and isinstance(y.owner.op, T.Alloc):
+            # `z` is the input of the Alloc op, i.e. T.alloc(z, <shape>)
+            z = y.owner.inputs[0]
+
+            # The following check is to prevent removal of shape errors.
+            try:
+                same_shape = node.fgraph.shape_feature.same_shape
+
+                # Get the subtensor of `x` indexed by `i` in order to compare
+                # shapes later.
+                if isinstance(node.op, IncSubtensor):
+                    xi = Subtensor(node.op.idx_list)(x, *i)
+                elif isinstance(node.op, AdvancedIncSubtensor):
+                    xi = advanced_subtensor(x, *i)
+                elif isinstance(node.op, AdvancedIncSubtensor1):
+                    xi = advanced_subtensor1(x, *i)
+                else:
+                    raise Exception('Should never happen!')
+
+                # Add `xi` to the `fgraph`. This is important for shape
+                # inference later because the variable must be part of the
+                # function graph in order to call `same_shape` on it.
+                reason = 'local_useless_incsubtensor_alloc'
+                node.fgraph.__import_r__([xi], '%s: add `xi`' % reason)
```

Comment:
Importing it in the fgraph itself isn't a good idea. This have some side error. You would need to remove it from the fggraph later. I think a better way is it include it just in the shape_feature. This can be done like this:

```
shape_feature.on_import(node.fgraph, xi.owner, reason)
```

This do not ask for clean up later. Idem for the later __import_r__